phoenix-core/src/main/java/org/apache/phoenix/jdbc/PhoenixConnection.java

```diff
@@ -441,6 +453,17 @@ private static Properties filterKnownNonProperties(Properties info) {
         return result.build();
     }
 
+    public boolean isInternalConnection() {
+        return isInternalConnection;
+    }
+
+    public void addChildConnection(PhoenixConnection connection) {
```

Comment:
If you're worried about multiple threads calling `addChildConnection` on the same connection object, then this method is still not thread safe. You need to synchronize the null check and element addition as an entire atomic operation.

Reply:
And then, if we end up synchronizing over all accesses of the list anyways, we can probably do with a normal ArrayList too

Reply:
I'm hoping to only synchronize of the construction of the iterable, as I hope most threads will be able to add themselves in parallel as this is a possible new slowdown, think the effort is worth it over normal ArrayList?